toolRepository-remote/src/main/scala/com/codacy/toolRepository/remote/ToolRepositoryRemote.scala

```diff
@@ -1,12 +1,105 @@
 package com.codacy.toolRepository.remote
 
-import com.codacy.analysis.core.model.{PatternSpec, ToolSpec}
+import akka.stream.Materializer
+import akka.stream.scaladsl.Sink
+import com.codacy.analysis.clientapi.definitions
+import com.codacy.analysis.clientapi.definitions.{PaginationInfo, PatternListResponse, ToolListResponse}
+import com.codacy.analysis.clientapi.tools.{ListToolPatternsResponse, ListToolsResponse, ToolsClient}
+import com.codacy.analysis.core.model.{ParameterSpec, PatternSpec, ToolSpec}
 import com.codacy.analysis.core.tools.ToolRepository
+import com.codacy.plugins.api.languages.Languages
 
-//TODO: Implement me
-class ToolRepositoryRemote() extends ToolRepository {
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext}
 
-  override def list(): Seq[ToolSpec] = ???
+class ToolRepositoryRemote(toolsClient: ToolsClient)(implicit val ec: ExecutionContext, implicit val mat: Materializer)
+    extends ToolRepository {
 
-  override def listPatterns(toolUuid: String): Seq[PatternSpec] = ???
+  override def list(): Seq[ToolSpec] = {
+    val source = PaginatedApiSourceFactory { cursor =>
+      toolsClient.listTools(cursor).value.map {
+        case Right(ListToolsResponse.OK(ToolListResponse(None | Some(PaginationInfo(None, _, _)), data))) =>
+          (None, data)
+        case Right(ListToolsResponse.OK(ToolListResponse(Some(PaginationInfo(newCursor: Some[String], _, _)), data))) =>
+          (newCursor, data)
+        case Right(ListToolsResponse.BadRequest(badRequest)) =>
+          throw new Exception(s"Failed to list tools with a Bad Request: ${badRequest.message}")
+        case Right(ListToolsResponse.InternalServerError(internalServerError)) =>
+          throw new Exception(s"Failed to list tools with an Internal Server Error: ${internalServerError.message}")
+        case Left(Right(error)) => throw new Exception(s"Failed to list tools with ${error.status} status code")
+        case Left(Left(throwable)) =>
+          throw new Exception(s"Failed to list tools failed: ${throwable.getMessage}", throwable)
+      }
+    }
+
+    val toolsF = source.runWith(Sink.seq).map(_.map(toToolSpec))
+    Await.result(toolsF, 1 minute)
```

Comment:
Is it possible to remove this `Await`?

Reply:
It is. @machadoit and I decided to try it first without futures, but we should revisit that before closing the issue

Reply:
But we need to have an await somewhere. Although we wish it would be closer to the top layer